### PR TITLE
ci: run on 12,14,16,18 node versions and run tests once linting is success

### PR DESCRIPTION
**Pull request by saihaj**
Just like [`graphql-yoga`](https://github.com/dotansimha/graphql-yoga) we should probably ensure this on all the Node versions we support. 

To avoid running too many CI tasks ensure that linting and prettier passes. Since if they fail we wouldn't merge PR anyways so no point running tests since user will have to push again.

### CODE REVIEW

**Review by B2o5T — APPROVED** [review]

**Comment by Urigo** [comment]
could/should we share parts of `.github/workflows/main.yml` between our libraries?

**Comment by saihaj** [comment]
> could/should we share parts of `.github/workflows/main.yml` between our libraries?

Yeah we should create composite actions I think with new bob we will be able to simplify more scripts cc @n1ru4l

**Review by ardatan — APPROVED** [review]

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -108,10 +111,15 @@ jobs:
...
+    needs:
+      - lint
+      - prettier-check
+      - dev-tests
+      - esm
     strategy:
       matrix:
         os: [ubuntu-latest] # remove windows to speed up the tests
-        node_version: [12, 17]
+        node_version: [12, 14, 16, 18]
```
Comment (B2o5T) [inline .github/workflows/main.yml:122]: I guess you can already remove 12
```suggestion
        node_version: [14, 16, 18]
```

**Review by B2o5T — APPROVED** [review]

File: .github/workflows/main.yml (hunk 1 above)
Comment (saihaj) [reply .github/workflows/main.yml:122]: we still support `v12` I think?

**Pull request merged**